ingestion_tools/scripts/data_validation/tests/test_frame.py

```diff
@@ -0,0 +1,128 @@
+                # this loop takes a long time, need to multithread
+                assert all(page.shape == curr_dimensions for page in header), "Not all pages have the same dimensions"
+                if frames_dimensions is None:
+                    frames_dimensions = curr_dimensions
+                elif curr_dimensions != frames_dimensions:
+                    errors.append(f"Frame dimensions do not match: {curr_dimensions} != {frames_dimensions}")
+            elif isinstance(header, MrcInterpreter):
+                if frames_dimensions is None:
+                    frames_dimensions = (header.nx, header.ny)
+                elif (header.nx, header.ny) != frames_dimensions:
+                    errors.append(f"Frame dimensions do not match: ({header.nx}, {header.ny}) != {frames_dimensions}")
+
+        assert not errors, "\n".join(errors)
+
+    ### END Self-consistency tests ###
+
+    ### BEGIN Tiltseries consistency tests ###
+    def test_frames_count(self, frames_files: List[str], tiltseries_metadata: Dict):
+        """Check that the number of frames is consistent between the metadata and the frame files."""
+        assert len(frames_files) == tiltseries_metadata["frames_count"]
+
+    def check_angles(
+        self,
+        angles: np.ndarray,
+        frames_dir: str,
+        frames_files: List[str],
+        # tiltseries_metadata: Dict,
+    ):
+        tilt_angle_to_frame_mapping: Dict[Union[str, float], str] = {}  # only str key for -0.0
+        remaining_frames_files = frames_files.copy()
+
+        for angle in angles:
+            # Look for a frame file that contains the tilt angle
+            angle_str = "({:.1f}|{:.2f})".format(angle, angle).replace(".", r"\.")
+            angle_regex = re.compile(f"{frames_dir}.*_{angle_str}.*")
+            angle_files = list(filter(angle_regex.match, remaining_frames_files))
+            assert len(angle_files) <= 1
+            if len(angle_files) == 0:
+                warnings.warn(f"Missing frame file for tilt angle {angle}", stacklevel=2)
+                continue
+            remaining_frames_files.remove(angle_files[0])
+            if angle in tilt_angle_to_frame_mapping:
+                warnings.warn(f"Duplicate / ambiguous frame file for tilt angle {angle}", stacklevel=2)
+            tilt_angle_to_frame_mapping[angle] = angle_files[0]
+
+        # special case for -0.0
+        angle_str = "(-0.0|-0.00)"
+        angle_regex = re.compile(f"{frames_dir}.*_{angle_str}.*")
+        angle_files = list(filter(angle_regex.match, remaining_frames_files))
+        assert len(angle_files) <= 1
+        if len(angle_files) == 1:
+            remaining_frames_files.remove(angle_files[0])
+            # usually key is float, but -0.0 is a special case
+            tilt_angle_to_frame_mapping["-0.0"] = angle_files[0]
+
+        assert remaining_frames_files == [], f"Frame files not accounted for: {remaining_frames_files}"
+
```

Comment:
A somewhat janky way to verify that the angles are correct, something to discuss with Utz